parcels/interaction/hash_flat.py

```diff
@@ -0,0 +1,145 @@
+            return False
+        if new_active_mask is None:
+            active_values = new_values
+        else:
+            active_values = new_values[:, new_active_mask]
+        for i_dim in range(3):
+            if np.any(active_values[i_dim, :] - self._box[i_dim][0] < 0):
+                return False
+            if np.any(active_values[i_dim, :] - self._box[i_dim][1] > 0):
+                return False
+        return True
+
+    def rebuild(self, values, active_mask=-1):
+        super().rebuild(values, active_mask)
+        active_values = self._values[:, self._active_mask]
+
+        # Compute the dimensions of the box with a margin.
+        self._box = []
+        for i_dim in range(3):
+            val_min = active_values[i_dim, :].min()
+            val_max = active_values[i_dim, :].max()
+            margin = (val_max-val_min)*0.3
+            self._box.append([val_min-margin, val_max+margin])
+
+        self._box = np.array(self._box)
+
+        epsilon = 1e-8
+
+        # Compute hash tables and the bits to compute it.
+        n_bits = ((self._box[:, 1] - self._box[:, 0]
+                   )/self.inter_dist.reshape(-1) + epsilon)/np.log(2)
+        self._bits = np.ceil(n_bits).astype(int)
+        self._min_box = self._box[:, 0]
+        self._min_box = self._min_box.reshape(-1, 1)
+        particle_hashes = self._values_to_hashes(values, self.active_idx)
+        self._hashtable = hash_split(particle_hashes,
+                                     active_idx=self.active_idx)
+        self._particle_hashes = particle_hashes
+
+        # Keep track of the position of a particle index within a cell.
+        self._hash_idx = np.empty_like(self._particle_hashes, dtype=int)
+        for idx_array in self._hashtable.values():
+            self._hash_idx[idx_array] = np.arange(len(idx_array))
+
+    def _values_to_hashes(self, values, active_idx=None):
+        if active_idx is None:
+            active_values = values
+        else:
+            active_values = values[:, active_idx]
+
+        # Compute the box_id/hashes.
+        box_i = ((active_values-self._min_box)/self.inter_dist).astype(int)
+        particle_hashes = np.bitwise_or(
+            box_i[0, :], np.left_shift(box_i[1, :], self._bits[0]))
+
+        if active_values is None:
+            return particle_hashes
+
+        # Put the hashes back
+        all_hashes = np.empty(values.shape[1], dtype=int)
+        all_hashes[active_idx] = particle_hashes
+        return all_hashes
+
+
+# @njit
+def hash_to_neighbors(hash_id, bits):
+    """Compute neighboring cells from a hash.
+
+    :param hash_id: hash value of the current cell.
+    :param bits: key to compute the hashesh.
+    :returns neighbors: List of cells neighboring hash_id.
+    """
+    coor = np.zeros((len(bits),), dtype=np.int32)
+    new_coor = np.zeros((len(bits),), dtype=np.int32)
+#     coor = np.zeros((len(bits),), dtype=nb.int32)
+#     new_coor = np.zeros((len(bits),), dtype=nb.int32)
```

Comment:
Bumba stray code-comments - please remove the commented lines.

Reply:
Done